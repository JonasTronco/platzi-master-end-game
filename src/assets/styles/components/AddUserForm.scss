@import '../main.scss';

.add-user {
  @include card;
  padding: $space_x6;
  &__type {
    margin-bottom: $space_x6;
    &__select {
      width: 150px;
    }
  }
  &__input {
    &__field {
      width: 250px;
      height: 30px;
      margin-bottom: $space_x3;
      margin-right: $space_x6;
    }
  }
  &__title-section {
    margin-bottom: $space_x3;
  }
  &__information-container {
    display: flex;
    flex-wrap: wrap;
    flex-direction: row;
    margin-bottom: $space_x3;
  }
  &__confirm {
    width: 100%;
    height: 30px;
    display: flex;
    justify-content: flex-end;
    &__button {
      width: 250px;
    }
  }
}
