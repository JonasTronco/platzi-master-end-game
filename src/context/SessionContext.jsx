import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
