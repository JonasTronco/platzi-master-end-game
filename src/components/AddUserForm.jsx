import { useState } from 'react';
import Router from 'next/router';
import axios from 'axios';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { useSession } from '../context/SessionContext';
import '../assets/styles/components/AddUserForm.scss';
import ErrorIcon from '../assets/icons/error.svg';
import SuccessIcon from '../assets/icons/success.svg';
import Modal from './Modal';

function AddUserForm() {
  const { session } = useSession();
  const [prevImg, setPrevImg] = useState(
    'https://res.cloudinary.com/alexisaraujoapp1/image/upload/v1590279706/alexisapp/ou8drr0g7poddnoqprmb.jpg',
  );
  const [errorImg, setErrorImg] = useState(null);
  const [response, setResponse] = useState({});
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  const [openModal, setOpenModal] = useState(false);

  async function createUser(userData) {
    setOpenModal(true);
    setLoading(true);
    try {
      const URL = `${process.env.NEXT_PUBLIC_API_URL}/users`;
      const config = {
        headers: { Authorization: `Bearer ${session.token}` },
      };
      const res = await axios.post(URL, userData, config);
      setResponse(res.data);
    } catch (err) {
      setError(err.data);
    }
    setLoading(false);
  }

  async function uploadImg() {
    setErrorImg('Loading image...');
    const imagedata = document.querySelector('input[type="file"]').files[0];
    if (imagedata.type.includes('image')) {
      const formData = new FormData();
      formData.append('file', imagedata);
      formData.append('upload_preset', 'alexisapp');
      const res = await axios.post(
        'https://api.cloudinary.com/v1_1/alexisaraujoapp1/image/upload',
        formData,
      );
      const { data } = res;
      setPrevImg(data.secure_url);
      setErrorImg(null);
    } else {
      setErrorImg('The selected file is not an image');
    }
  }

  function closeModal() {
    setOpenModal(false);
  }

  function AddMessage() {
    if (loading) {
      return (
        <div className="message">
          <div className="loader" />
        </div>
      );
    }

    if (error) {
      return (
        <div className="message">
          <ErrorIcon className="message__icon--negative" />
          <strong className="message__text">{error.message}</strong>
          <button className="btn" type="button" onClick={closeModal}>
            Accept
          </button>
        </div>
      );
    }

    if (response) {
      return (
        <div className="message">
          <SuccessIcon className="message__icon--positive" />
          <strong className="message__text">{response.message}</strong>
          <button
            className="btn"
            type="button"
            onClick={() => {
              Router.push(`/detail/${response.data._id}`);
            }}
          >
            Accept
          </button>
        </div>
      );
    }

    return '';
  }

  function AvatarForm() {
    return (
      <form className="avatar-form">
        {prevImg ? (
          <figure className="avatar-form__preview">
            <img
              className="avatar-form__preview__img"
              src={prevImg}
              alt="Avatar"
            />
          </figure>
        ) : (
          ''
        )}
        <div className="input">
          <label className="input__label">
            Profile Image
            <input
              className="input__field--img"
              type="file"
              onChange={uploadImg}
            />
          </label>
          {errorImg ? <span className="input__error">{errorImg}</span> : ''}
        </div>
      </form>
    );
  }

  function UserForm() {
    return (
      <Formik
        initialValues={{
          imageURL: prevImg,
          typeOfUser: 'Patient',
          firstName: '',
          lastName: '',
          documentID: '',
          email: '',
          contactNumber: '',
        }}
        validate={(values) => {
          const errors = {};
          if (!values.typeOfUser) {
            errors.typeOfUser = 'Please choose a type';
          }
          if (!values.firstName) {
            errors.firstName = 'Please enter a first name';
          }
          if (!values.lastName) {
            errors.lastName = 'Please enter a last name';
          }
          if (!values.documentID) {
            errors.documentID = 'Please enter a document id';
          }
          if (!values.email) {
            errors.email = 'Please enter a email';
          } else if (
            !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
          ) {
            errors.email = 'Invalid email address';
          }
          if (!values.contactNumber) {
            errors.contactNumber = 'Please enter a phone number';
          }
          return errors;
        }}
        onSubmit={async (values, { setSubmitting }) => {
          await createUser(values);
          setSubmitting(false);
        }}
      >
        {({ isSubmitting }) => (
          <Form className="user-add-form">
            <section className="user-add-form__section">
              <div className="input">
                <label className="input__label">
                  Type
                  <Field
                    className="input__field"
                    as="select"
                    name="typeOfUser"
                    disabled={isSubmitting}
                  >
                    <option disabled="disabled">Choose a type</option>
                    <option value="Patient">Patient</option>
                    <option value="Doctor">Doctor</option>
                    <option value="Bacteriologist">Bacteriologist</option>
                    <option value="Administrator">Administrator</option>
                  </Field>
                </label>
                <ErrorMessage
                  name="typeOfUser"
                  component="span"
                  className="input__error"
                />
              </div>
            </section>
            <h2 className="user-add-form__title">User Information</h2>
            <section className="user-add-form__section">
              <div className="input">
                <label className="input__label">
                  First Name
                  <Field
                    className="input__field"
                    type="text"
                    name="firstName"
                    disabled={isSubmitting}
                  />
                </label>
                <ErrorMessage
                  name="firstName"
                  component="span"
                  className="input__error"
                />
              </div>
              <div className="input">
                <label className="input__label">
                  Last Name
                  <Field
                    className="input__field"
                    type="text"
                    name="lastName"
                    disabled={isSubmitting}
                  />
                </label>
                <ErrorMessage
                  name="lastName"
                  component="span"
                  className="input__error"
                />
              </div>
              <div className="input">
                <label className="input__label">
                  Document ID
                  <Field
                    className="input__field"
                    type="number"
                    name="documentID"
                    disabled={isSubmitting}
                  />
                </label>
                <ErrorMessage
                  name="documentID"
                  component="span"
                  className="input__error"
                />
              </div>
            </section>
            <h2 className="user-add-form__title">Contact</h2>
            <section className="user-add-form__section">
              <div className="input">
                <label className="input__label">
                  email
                  <Field
                    className="input__field"
                    type="email"
                    name="email"
                    disabled={isSubmitting}
                  />
                </label>
                <ErrorMessage
                  name="email"
                  component="span"
                  className="input__error"
                />
              </div>
              <div className="input">
                <label className="input__label">
                  Phone Number
                  <Field
                    className="input__field"
                    type="number"
                    name="contactNumber"
                    disabled={isSubmitting}
                  />
                </label>
                <ErrorMessage
                  name="contactNumber"
                  component="span"
                  className="input__error"
                />
              </div>
            </section>
            <button className="btn" type="submit" disabled={isSubmitting}>
              Add User
            </button>
            {loading ? (
              <span className="user-add-form__load">Creating user...</span>
            ) : (
              ''
            )}
          </Form>
        )}
      </Formik>
    );
  }

  return (
    <>
      <AvatarForm />
      <UserForm />
      <Modal isOpen={openModal} onClose={closeModal}>
        <AddMessage />
      </Modal>
    </>
  );
}
export default AddUserForm;
