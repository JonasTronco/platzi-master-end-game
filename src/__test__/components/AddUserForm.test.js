import AddUserForm from '../../components/AddUserForm';
import { create } from 'react-test-renderer';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { mount } from 'enzyme';
import * as nextRouter from 'next/router';
import { SessionProvider } from '../../context/SessionContext';
describe('AddUserForm Component', () => {
  nextRouter.useRouter = jest.fn();
  nextRouter.useRouter.mockImplementation(() => ({
    route: '/',
    query: {
      orderID: '5ecda76661b17e50c87721ab'
    }
  }));
  const obj = {
    session: {
      token: '',
      user: {
        id: '',
        username: '',
        typeOfUser: '',
        isActive: true,
        imageURL: '',
        firstName: '',
        lastName: '',
        defaultPath: ''
      }
    }
  };
  test('Should render AddUserForm component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <AddUserForm />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Should AddUserForm button renders fine', () => {
    const component = render(
      <SessionProvider value={obj}>
        <AddUserForm />
      </SessionProvider>
    );
    expect(component.getByTestId('button-submit')).toBeInTheDocument();
  });
  it('<AddUserForm />', () => {
    const component = mount(
      <SessionProvider value={obj}>
        <AddUserForm />
      </SessionProvider>
    );
    expect(component.length).toEqual(1);
  });
});
