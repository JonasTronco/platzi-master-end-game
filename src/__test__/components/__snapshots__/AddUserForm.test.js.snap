// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserForm Component Should render AddUserForm component 1`] = `
Array [
  <form
    className="avatar-form"
    data-testid="addAvatar"
  >
    <figure
      className="avatar-form__preview"
    >
      <img
        alt="Avatar"
        className="avatar-form__preview__img"
        src="https://res.cloudinary.com/alexisaraujoapp1/image/upload/v1590279706/alexisapp/ou8drr0g7poddnoqprmb.jpg"
      />
    </figure>
    <div
      className="input"
    >
      <label
        className="input__label"
      >
        Profile Image
        <input
          className="input__field--img"
          onChange={[Function]}
          type="file"
        />
      </label>
      
    </div>
  </form>,
  <form
    action="#"
    className="user-add-form"
    data-testid="add-form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <section
      className="user-add-form__section"
    >
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          Type
          <select
            className="input__field"
            disabled={false}
            name="typeOfUser"
            onBlur={[Function]}
            onChange={[Function]}
            value="Patient"
          >
            <option
              disabled="disabled"
            >
              Choose a type
            </option>
            <option
              value="Patient"
            >
              Patient
            </option>
            <option
              value="Doctor"
            >
              Doctor
            </option>
            <option
              value="Bacteriologist"
            >
              Bacteriologist
            </option>
            <option
              value="Administrator"
            >
              Administrator
            </option>
          </select>
        </label>
      </div>
    </section>
    <h2
      className="user-add-form__title"
    >
      User Information
    </h2>
    <section
      className="user-add-form__section"
    >
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          First Name
          <input
            className="input__field"
            disabled={false}
            name="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          Last Name
          <input
            className="input__field"
            disabled={false}
            name="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          Document ID
          <input
            className="input__field"
            disabled={false}
            name="documentID"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value=""
          />
        </label>
      </div>
    </section>
    <h2
      className="user-add-form__title"
    >
      Contact
    </h2>
    <section
      className="user-add-form__section"
    >
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          email
          <input
            className="input__field"
            disabled={false}
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="email"
            value=""
          />
        </label>
      </div>
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          Phone Number
          <input
            className="input__field"
            disabled={false}
            name="contactNumber"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value=""
          />
        </label>
      </div>
    </section>
    <button
      className="btn"
      data-testid="button-submit"
      disabled={false}
      type="submit"
    >
      Add User
    </button>
    
  </form>,
]
`;
